src/main/java/com/synopsys/integration/alert/web/security/authentication/AlertAuthenticationProvider.java

```diff
@@ -0,0 +1,115 @@
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.ldap.authentication.LdapAuthenticationProvider;
+import org.springframework.stereotype.Component;
+
+import com.synopsys.integration.alert.common.exception.AlertLDAPConfigurationException;
+import com.synopsys.integration.alert.common.rest.model.UserModel;
+import com.synopsys.integration.alert.database.user.UserRole;
+import com.synopsys.integration.alert.web.security.authentication.ldap.LdapManager;
+
+@Component
+public class AlertAuthenticationProvider implements AuthenticationProvider {
+    private static final Logger logger = LoggerFactory.getLogger(AlertAuthenticationProvider.class);
+    private final DaoAuthenticationProvider alertDatabaseAuthProvider;
+    private final LdapManager ldapManager;
+
+    @Autowired
+    public AlertAuthenticationProvider(final DaoAuthenticationProvider alertDatabaseAuthProvider, final LdapManager ldapManager) {
+        this.alertDatabaseAuthProvider = alertDatabaseAuthProvider;
+        this.ldapManager = ldapManager;
+    }
+
+    @Override
+    public Authentication authenticate(final Authentication authentication) throws AuthenticationException {
+        if (!supports(authentication.getClass())) {
+            throw new IllegalArgumentException("Only UsernamePasswordAuthenticationToken is supported, " + authentication.getClass() + " was attempted");
+        }
+        Authentication authenticationResult = performLdapAuthentication(authentication);
+        if (!authenticationResult.isAuthenticated()) {
+            authenticationResult = performDatabaseAuthentication(authentication);
+        }
+        final Collection<? extends GrantedAuthority> authorities = isAuthorized(authenticationResult) ? authenticationResult.getAuthorities() : List.of();
+        final UsernamePasswordAuthenticationToken authenticationToken = new UsernamePasswordAuthenticationToken(authenticationResult.getPrincipal(), authenticationResult.getCredentials(), authorities);
+        SecurityContextHolder.getContext().setAuthentication(authenticationToken);
+        return authenticationToken;
+    }
+
+    @Override
+    public boolean supports(final Class<?> authentication) {
+        return UsernamePasswordAuthenticationToken.class.isAssignableFrom(authentication);
+    }
+
+    private Authentication performDatabaseAuthentication(final Authentication pendingAuthentication) {
+        logger.info("Attempting database authentication...");
+        return alertDatabaseAuthProvider.authenticate(pendingAuthentication);
+    }
+
+    private Authentication performLdapAuthentication(final Authentication pendingAuthentication) {
+        logger.info("Checking ldap based authentication...");
+        Authentication result;
```

Comment:
We could set result = pendingAuthentication right here then only add the success one if it never throws the exception